Switch MinGW building to CMake (#1906)
gbdev/rgbds: CMakeLists.txt

include(CTest) # Note: CTest only functions properly if included from the top-level CMakeLists.
include(GNUInstallDirs)
include(CMakeDependentOption)

## Compiler switches that should apply to our deps too.

option(SANITIZERS "Build with sanitizers enabled" OFF)
cmake_dependent_option(MORE_WARNINGS "Turn on more warnings" OFF "NOT MSVC" OFF)

if(SANITIZERS)
  if(MSVC)
    message(STATUS "ASan enabled")
    add_compile_options(/fsanitize=address) # Note that this shouldn't be passed to the linker.

>   else(MSVC) # We assume GNU compilers.

Reviewer: Since when do we assume GNU compilers? We test with GCC (GNU) and Clang (GNU-compatible), but don't use any GNU-isms (`-std=c++20 -pedantic`).

Also I can't tell what the comment has to do with the "`if(MSVC) ... else(MSVC) ... endif()`" block. And from the [CMake `else` docs](https://cmake.org/cmake/help/latest/command/else.html):

> The optional `<condition>` argument is supported for backward compatibility only and is not evaluated.

Author: We assume GNU-style command lines and flags; for example, `-fstuff`. The comment is here to state “if MSVC then <handling>, else we assume a GNU-like compiler”. I can remove the repeated `MSVC`, though.

Reviewer: ```suggestion
  else() # If not MSVC, we assume support for GNU-like sanitizer flags.
```

Author: I rephrased that. What about now?

Reviewer: Sure.

    # be able to override these easily so we put them first.
    string(PREPEND CMAKE_CXX_FLAGS_DEBUG "-Og -fno-omit-frame-pointer -fno-optimize-sibling-calls ")
  endif()

> endif(SANITIZERS)

Reviewer: From the [CMake `endif` docs](https://cmake.org/cmake/help/latest/command/endif.html#command:endif):

> The optional `<condition>` argument is supported for backward compatibility only. If used it must be a verbatim repeat of the argument of the opening `if` clause.

We don't do this anywhere else, even for longer conditional blocks than this; and I wouldn't want to set a precedent (`endif(some long condition)` would be annoying, and the entire practice just reminds me intro-CS requiring `} // end myMethodName` comments everywhere).

```suggestion
endif()
```

  )
  add_definitions(/D_CRT_SECURE_NO_WARNINGS)
else()
  add_compile_options(-Wall -pedantic -fno-exceptions -fno-rtti -Wno-unknown-warning-option
                      # C++20 allows macros to take zero variadic arguments.
                      # Some versions of Clang don't recognize this, and treat them as a GNU extension.
                      -Wno-gnu-zero-variadic-macro-arguments)

  if(MORE_WARNINGS)
    add_compile_options(-Wextra
                        -Walloc-zero -Wcast-align -Wcast-qual -Wduplicated-branches -Wduplicated-cond
                        -Wfloat-equal -Wlogical-op -Wnull-dereference -Wold-style-cast -Wshift-overflow=2
                        -Wstringop-overflow=4 -Wtrampolines -Wundef -Wuninitialized -Wunused -Wshadow
                        -Wformat=2 -Wformat-overflow=2 -Wformat-truncation=1
                        -Wno-format-nonliteral -Wno-strict-overflow
                        -Wno-unused-but-set-variable # bison's `yynerrs_` is incremented but unused
                        -Wno-type-limits -Wno-tautological-constant-out-of-range-compare
                        -Wvla) # MSVC does not support VLAs
  endif()

  if(MINGW)

>     # MinGW warns about some format specifiers which are missing on Windows XP and earlier.

Reviewer: Mentioning "%z" and "%j" is more likely to remind people "oh yeah, I've seen the compiler complain about those before", and is greppable.
```suggestion
    # MinGW warns about format specifiers like "%z" and "%j" which are missing on Windows XP and earlier.
```

>     # we rely on runtimes more modern than that, so we can ignore those warnings.

Reviewer: ```suggestion
    # MinGW warns about format specifiers "%z" and "%j", which are missing on Windows XP and earlier.
    # We rely on runtimes more modern than that, so we can ignore those warnings.
```

set(CPACK_VERBATIM_VARIABLES ON)
set(CPACK_THREADS 0) # Use all available CPU cores.
set(CPACK_INCLUDE_TOPLEVEL_DIRECTORY OFF)

set(CPACK_INSTALL_CMAKE_PROJECTS "${CMAKE_BINARY_DIR};${CMAKE_PROJECT_NAME};runtime;/")
#set(CPACK_ZIP_COMPONENT_INSTALL ON)
#cpack_add_component(runtime)
#cpack_add_component(man)

> #set(CPACK_COMPONENTS_ALL "runtime;man")

Reviewer: If this works, uncomment it; if not, remove it. (Leftovers from an experiment to use the CPack archive generator for something?)

Author: lol yeah this is a remnant. I'll remove it; thanks.
